File: webapp/scripts/test-rafael-fix.ts
#!/usr/bin/env npx tsx
/**
 * Test script to verify the TFSA-First household rebalancing fix
 * Tests Rafael & Lucy scenario from juanclavierb@gmail.com
 */

const PYTHON_API_URL = 'http://localhost:8000';

// Rafael & Lucy's actual data from the database
const rafaelLucyScenario = {
  p1: {
    name: "Rafael",
    start_age: 67,
    cpp_start_age: 65,
    cpp_annual_at_start: 15500,
    oas_start_age: 65,
    oas_annual_at_start: 8500,
    employer_pension_annual: 0,
    rental_income_annual: 0,
    other_income_annual: 0,
    tfsa_balance: 312000,
    rrif_balance: 350000,
    rrsp_balance: 0,
    nonreg_balance: 915000,
    corporate_balance: 0,
    nonreg_acb: 915000,
    nr_cash: 91500,
    nr_gic: 183000,
    nr_invest: 640500,
    y_nr_cash_interest: 2.0,
    y_nr_gic_interest: 3.5,
    y_nr_inv_total_return: 6.0,
    y_nr_inv_elig_div: 2.0,
    y_nr_inv_nonelig_div: 0.5,
    y_nr_inv_capg: 3.0,
    y_nr_inv_roc_pct: 0.5,
    nr_cash_pct: 10.0,
    nr_gic_pct: 20.0,
    nr_invest_pct: 70.0,
    corp_cash_bucket: 0,
    corp_gic_bucket: 0,
    corp_invest_bucket: 0,
    corp_rdtoh: 0,
    y_corp_cash_interest: 2.0,
    y_corp_gic_interest: 3.5,
    y_corp_inv_total_return: 6.0,
    y_corp_inv_elig_div: 2.0,
    y_corp_inv_capg: 3.5,
    corp_cash_pct: 5.0,
    corp_gic_pct: 10.0,
    corp_invest_pct: 85.0,
    corp_dividend_type: "eligible" as const,
    tfsa_room_start: 7000,
    tfsa_contribution_annual: 0,
  },
  p2: {
    name: "Lucy",
    start_age: 70,
    cpp_start_age: 65,
    cpp_annual_at_start: 15500,
    oas_start_age: 65,
    oas_annual_at_start: 8500,
    employer_pension_annual: 0,
    rental_income_annual: 0,
    other_income_annual: 0,
    tfsa_balance: 114000,
    rrif_balance: 22000,
    rrsp_balance: 0,
    nonreg_balance: 183000,
    corporate_balance: 0,
    nonreg_acb: 183000,
    nr_cash: 18300,
    nr_gic: 36600,
    nr_invest: 128100,
    y_nr_cash_interest: 2.0,
    y_nr_gic_interest: 3.5,
    y_nr_inv_total_return: 6.0,
    y_nr_inv_elig_div: 2.0,
    y_nr_inv_nonelig_div: 0.5,
    y_nr_inv_capg: 3.0,
    y_nr_inv_roc_pct: 0.5,
    nr_cash_pct: 10.0,
    nr_gic_pct: 20.0,
    nr_invest_pct: 70.0,
    corp_cash_bucket: 0,
    corp_gic_bucket: 0,
    corp_invest_bucket: 0,
    corp_rdtoh: 0,
    y_corp_cash_interest: 2.0,
    y_corp_gic_interest: 3.5,
    y_corp_inv_total_return: 6.0,
    y_corp_inv_elig_div: 2.0,
    y_corp_inv_capg: 3.5,
    corp_cash_pct: 5.0,
    corp_gic_pct: 10.0,
    corp_invest_pct: 85.0,
    corp_dividend_type: "eligible" as const,
    tfsa_room_start: 7000,
    tfsa_contribution_annual: 0,
  },
  province: "AB",
  start_year: 2025,
  end_age: 95,
  strategy: "tfsa-first",
  spending_go_go: 160000,
  go_go_end_age: 75,
  spending_slow_go: 120000,
  slow_go_end_age: 85,
  spending_no_go: 90000,
  spending_inflation: 2.0,
  general_inflation: 2.0,
  tfsa_room_annual_growth: 7000,
  gap_tolerance: 1000,
  reinvest_nonreg_dist: false,
  income_split_rrif_fraction: 0.0,
  hybrid_rrif_topup_per_person: 0,
  stop_on_fail: false,
};

async function testRafaelScenario() {
  console.log('\n🧪 Testing Rafael & Lucy TFSA-First Strategy Fix\n');
  console.log('=' .repeat(70));

  try {
    const response = await fetch(`${PYTHON_API_URL}/api/run-simulation`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(rafaelLucyScenario),
    });

    if (!response.ok) {
      console.error(`❌ HTTP Error: ${response.status} ${response.statusText}`);
      const errorData = await response.text();
      console.error('Response:', errorData);
      return;
    }

    const data = await response.json();

    if (!data.success) {
      console.error('❌ Simulation failed:', data.message);
      if (data.errors) {
        console.error('Errors:', data.errors);
      }
      return;
    }

    console.log('✅ Simulation completed successfully!\n');

    // Check years 2032-2034 for gaps (the problematic years)
    const yearByYear = data.year_by_year || [];
    const problematicYears = [2032, 2033, 2034];

    console.log('📊 Checking Years 2032-2034 (Previously Had Gaps):\n');
    console.log('-'.repeat(70));

    let foundGaps = false;

    for (const targetYear of problematicYears) {
      const yearData = yearByYear.find((y: any) => y.year === targetYear);

      if (yearData) {
        const gap = yearData.spending_gap || 0;
        const tfsaP1 = yearData.tfsa_balance_p1 || 0;
        const tfsaP2 = yearData.tfsa_balance_p2 || 0;
        const tfsaWithdrawalP1 = yearData.tfsa_withdrawal_p1 || 0;
        const tfsaWithdrawalP2 = yearData.tfsa_withdrawal_p2 || 0;
        const totalTax = yearData.total_tax || 0;

        const hasGap = Math.abs(gap) > 1000; // Allowing $1k tolerance
        if (hasGap) foundGaps = true;

        const status = hasGap ? '❌ GAP FOUND' : '✅ FUNDED';

        console.log(`\nYear ${targetYear}: ${status}`);
        console.log(`  Spending Gap:        $${gap.toLocaleString()}`);
        console.log(`  TFSA Balance P1:     $${tfsaP1.toLocaleString()}`);
        console.log(`  TFSA Balance P2:     $${tfsaP2.toLocaleString()}`);
        console.log(`  TFSA Withdrawn P1:   $${tfsaWithdrawalP1.toLocaleString()}`);
        console.log(`  TFSA Withdrawn P2:   $${tfsaWithdrawalP2.toLocaleString()}`);
        console.log(`  Total Tax Paid:      $${totalTax.toLocaleString()}`);
      } else {
        console.log(`\nYear ${targetYear}: ⚠️  Data not found`);
      }
    }

    console.log('\n' + '-'.repeat(70));

    // Summary
    const summary = data.summary || {};
    console.log('\n📈 Overall Simulation Summary:\n');
    console.log(`  Strategy:            ${rafaelLucyScenario.strategy.toUpperCase()}`);
    console.log(`  Years Simulated:     ${summary.years_simulated || 'N/A'}`);
    console.log(`  Years Funded:        ${summary.years_funded || 'N/A'}`);
    console.log(`  Success Rate:        ${(summary.success_rate || 0).toFixed(1)}%`);
    console.log(`  Total Tax Paid:      $${(summary.total_tax_paid || 0).toLocaleString()}`);
    console.log(`  Final Estate:        $${(summary.final_estate_after_tax || 0).toLocaleString()}`);
    console.log(`  Health Score:        ${summary.health_score || 'N/A'}/100 (${summary.health_rating || 'N/A'})`);

    console.log('\n' + '='.repeat(70));

    if (foundGaps) {
      console.log('\n❌ TEST FAILED: Gaps still exist in years 2032-2034');
      console.log('The household rebalancing fix may not be working correctly.\n');
      process.exit(1);
    } else {
      console.log('\n✅ TEST PASSED: No gaps found! The fix is working correctly.');
      console.log('TFSA withdrawals are now properly covering spending needs.\n');
      process.exit(0);
    }

  } catch (error) {
    console.error('❌ Error running simulation:', error);
    process.exit(1);
  }
}

testRafaelScenario();

File: test-rafael-fix.py
#!/usr/bin/env python3
"""
Test Rafael's scenario to verify the RRIF-Frontload fix
Rafael: Single, 60, retiring at 67, $400k assets (RRIF $350k, TFSA $50k)
Needs $120k/year spending, CPP/OAS at 65
Problem: Shows "Gap" in early years despite having sufficient assets
"""

import requests
import json
import sys
from datetime import datetime

# API endpoint (local development)
API_URL = "http://localhost:8000/api/run-simulation"

def test_rafael_scenario():
    """Test Rafael's exact scenario with RRIF-Frontload strategy"""

    # Rafael's data (single person, RRIF-heavy portfolio)
    test_data = {
        "p1": {
            "name": "Rafael",
            "is_retired": False,
            "retirement_age": 67,
            "start_age": 60,
            "life_expectancy": 85,

            # Assets - RRIF-heavy portfolio
            "rrsp_balance": 0,
            "rrif_balance": 350000,
            "tfsa_balance": 50000,
            "nonreg_balance": 0,
            "corporate_balance": 0,

            # Income settings - CPP/OAS at 65 (early)
            "cpp_start_age": 65,
            "cpp_amount": 8789,
            "oas_start_age": 65,
            "oas_amount": 6544,
            "gis_amount": 0,

            # Growth rates
            "yield_rrsp_growth": 0.04,
            "yield_rrif_growth": 0.04,
            "yield_tfsa_growth": 0.04,
            "yield_nonreg_growth": 0.04,

            # Other
            "tfsa_room_start": 0,
            "tfsa_room_annual_growth": 7000,
            "nonreg_acb": 0,
            "pension_incomes": [],
            "other_incomes": []
        },

        # Empty p2 for single person (API requires it but with proper structure)
        "p2": {
            "name": "",
            "is_retired": False,
            "retirement_age": 65,
            "start_age": 60,
            "life_expectancy": 85,
            "rrsp_balance": 0,
            "rrif_balance": 0,
            "tfsa_balance": 0,
            "nonreg_balance": 0,
            "corporate_balance": 0,
            "cpp_start_age": 65,
            "cpp_amount": 0,
            "oas_start_age": 65,
            "oas_amount": 0,
            "gis_amount": 0,
            "yield_rrsp_growth": 0.04,
            "yield_rrif_growth": 0.04,
            "yield_tfsa_growth": 0.04,
            "yield_nonreg_growth": 0.04,
            "tfsa_room_start": 0,
            "tfsa_room_annual_growth": 0,
            "nonreg_acb": 0,
            "pension_incomes": [],
            "other_incomes": []
        },

        # Household settings
        "household_is_couple": False,
        "withdrawal_strategy": "RRIF-Frontload (15%/8%)",  # RRIF-heavy strategy
        "spending_go_go": 120000,  # High spending need
        "spending_slow_go": 80000,
        "spending_no_go": 50000,
        "go_go_end_age": 75,
        "slow_go_end_age": 85,
        "end_age": 85,
        "gap_tolerance": 100,
        "stop_on_fail": False,
        "general_inflation": 0.025,
        "spending_inflation": 0.025,
        "province": "ON",
        "start_year": 2025
    }

    print("\n" + "="*80)
    print("TESTING RAFAEL'S SCENARIO - RRIF-FRONTLOAD FIX")
    print("="*80)
    print(f"Strategy: {test_data['withdrawal_strategy']}")
    print(f"Assets: RRIF=${test_data['p1']['rrif_balance']:,}, TFSA=${test_data['p1']['tfsa_balance']:,}")
    print(f"Spending: ${test_data['spending_go_go']:,}/year (Go-Go phase)")
    print(f"CPP/OAS: Starting at age {test_data['p1']['cpp_start_age']}")

    try:
        # Run simulation
        print("\nRunning simulation...")
        response = requests.post(API_URL, json=test_data, timeout=30)

        if response.status_code == 200:
            result = response.json()

            # Check first few years (where the bug was showing "Gap")
            print("\n" + "="*80)
            print("YEAR-BY-YEAR RESULTS (First 5 Years of Retirement)")
            print("="*80)

            retirement_start_year = 2025 + (67 - 60)  # 2032
            years_to_check = []

            for year_data in result.get("years", []):
                year = year_data.get("year", 0)
                if year >= retirement_start_year and year <= retirement_start_year + 4:
                    years_to_check.append(year_data)

            # Analyze results
            all_ok = True
            for year_data in years_to_check:
                year = year_data.get("year", 0)
                age = year_data.get("age_p1", 0)

                # Extract key values
                rrif_withdrawal = year_data.get("withdraw_rrif_p1", 0)
                tfsa_withdrawal = year_data.get("withdraw_tfsa_p1", 0)
                total_withdrawals = year_data.get("total_withdrawals", 0)

                # Calculate total income
                cpp = year_data.get("cpp_p1", 0)
                oas = year_data.get("oas_p1", 0)
                gis = year_data.get("gis_p1", 0)

                # Tax
                tax = year_data.get("total_tax_after_split", 0)

                # Net income
                gross_income = cpp + oas + gis + total_withdrawals
                net_income = gross_income - tax

                # Spending and gap
                spending = year_data.get("spend_target_after_tax", 0)
                gap = year_data.get("spending_gap", 0)
                is_underfunded = year_data.get("is_underfunded", False)

                # Remaining balances
                end_rrif = year_data.get("end_rrif_p1", 0)
                end_tfsa = year_data.get("end_tfsa_p1", 0)
                total_assets = end_rrif + end_tfsa

                # Status
                status = "Gap" if is_underfunded else "OK"
                if is_underfunded:
                    all_ok = False

                print(f"\nYear {year} (Age {age}):")
                print(f"  Withdrawals: RRIF=${rrif_withdrawal:,.0f}, TFSA=${tfsa_withdrawal:,.0f}, Total=${total_withdrawals:,.0f}")
                print(f"  Gov Benefits: CPP=${cpp:,.0f}, OAS=${oas:,.0f}, GIS=${gis:,.0f}")
                print(f"  Gross Income: ${gross_income:,.0f}")
                print(f"  Tax: ${tax:,.0f}")
                print(f"  Net Income: ${net_income:,.0f}")
                print(f"  Spending Need: ${spending:,.0f}")
                print(f"  Gap: ${gap:,.0f}")
                print(f"  Status: {status} {'❌' if is_underfunded else '✅'}")
                print(f"  Remaining Assets: RRIF=${end_rrif:,.0f}, TFSA=${end_tfsa:,.0f}, Total=${total_assets:,.0f}")

            # Summary
            print("\n" + "="*80)
            print("TEST RESULTS SUMMARY")
            print("="*80)

            if all_ok:
                print("✅ SUCCESS: No 'Gap' status in early retirement years!")
                print("The RRIF-Frontload fix is working correctly.")
                print("Rafael's simulation now properly withdraws sufficient funds to meet spending needs.")
            else:
                print("❌ FAILURE: Still seeing 'Gap' status in early years")
                print("The fix may not be fully addressing the issue.")

                # Additional debugging
                print("\n⚠️  DEBUGGING INFO:")
                print("Check if RRIF withdrawals are sufficient to meet after-tax spending needs.")
                print("With $120k spending and ~30% tax rate, need ~$170k gross withdrawals.")

            # Check health score
            health_score = result.get("health_score", 0)
            print(f"\nHealth Score: {health_score}/100")

            return all_ok

        else:
            print(f"\n❌ API Error: {response.status_code}")
            print(f"Response: {response.text}")
            return False

    except requests.exceptions.RequestException as e:
        print(f"\n❌ Connection Error: {e}")
        print("Make sure the API server is running on http://localhost:8000")
        return False
    except Exception as e:
        print(f"\n❌ Unexpected Error: {e}")
        return False

if __name__ == "__main__":
    success = test_rafael_scenario()
    sys.exit(0 if success else 1)